docs: Clean up GoCD alert script docs

timmc-edx opened the pull request:
- Link to integration that these need to be coordinated with
- Clarify what the auth token is
- Use "JSM" instead of "Opsgenie" everywhere (except to note the old name)
- Fix docstring of close-alert (which had been a copy of open-alert)
- Include alias in open-alert logging
- Add comment next to alias var explaining why it has to be of that form

doctoryes commented on tubular/utils/retry.py:
```diff
@@ -0,0 +1,145 @@
+
+def retry(attempts=MAX_ATTEMPTS, delay_seconds=DELAY_SECONDS, max_time_seconds=MAX_TIME_SECONDS):
+    """
+    Decorator wraps a function that will attempt to "retry" the function if an exception is raised during exeuction.
```
Nit: s/exeuction/execution.

macdiesel replied:
done

doctoryes commented on tubular/utils/retry.py:
```diff
@@ -0,0 +1,145 @@
+        if delay_seconds < 0:
+            raise RetryException(
+                "Must specify a delay_seconds number greater than 0. Value: {0}".format(delay_seconds))
+
```
Consider `max_attempts` = 3, `delay_seconds` = 5, and `max_time_seconds` = 4. In this case, no retries will be attempted. Do we want to warn the user in this case?

macdiesel replied:
done I have added a log.warn statement

doctoryes commented on tubular/utils/retry.py:
```diff
@@ -0,0 +1,145 @@
+            bool: True if the retry has reached it's maximum number of attempts
+                  False if the current attempt is less than the number of attempts
+        """
+        if self._current_attempt_number <= self.max_attempts:
```
Just `return self._current_attempt_number <= self.max_attempts`  ?

macdiesel replied:
solid advice, done.

doctoryes commented on tubular/utils/retry.py:
```diff
@@ -0,0 +1,145 @@
+            bool: True if the maximum runtime of the retry has been met or exceeded
+                  False if the max_time was not set or if the max time has not yet been reached
+        """
+        if self._max_datetime and self._max_datetime - datetime.utcnow() < timedelta(0, 0, 0):
```
Remove the `if` in favor of a single `return`?

macdiesel replied:
solid, done.

doctoryes commented on tubular/utils/retry.py:
```diff
@@ -0,0 +1,145 @@
+        Returns:
+            int: seconds to delay
+        """
+        return round(self.delay_seconds)
```
Maybe just `round` this in the `__init__`?

macdiesel replied:
done.

doctoryes commented on tubular/utils/retry.py:
```diff
@@ -0,0 +1,145 @@
+    def max_attempts_reached(self):
+        """
+        Returns:
+            bool: True if the retry has reached it's maximum number of attempts
```
s/it's/its

macdiesel replied:
done

doctoryes commented on tubular/utils/retry.py:
```diff
@@ -0,0 +1,145 @@
+            kwargs(dict<str:any>): Keyword arguments to the wrapped function
+
+        """
+        while not self.max_attempts_reached() and not self.max_time_reached():
```
Maybe pull this logic into a local method?:
```
def done():
    return self.max_attempts_reached() or self.max_time_reached()

while not done():
    ...
```

macdiesel replied:
done(), now with more testing

doctoryes commented on tubular/tests/test_retry.py:
```diff
@@ -0,0 +1,125 @@
+
+from ..utils import retry
+
+sys._is_retry_enabled = True
```
Why put this global var in `sys` versus in `utils/retry.py`? Seems like it belongs there.

feanil replied:
Why not an environment variable, we could have it disabled by default and only turn it on during the testing of the feature and in production?

doctoryes commented on tubular/tests/test_retry.py:
```diff
@@ -0,0 +1,125 @@
+        manager._current_attempt_number += 1
+        self.assertFalse(manager.max_attempts_reached())
+
+        # incrementing again should put us past the max attempds threshold
```
`attempts`

macdiesel replied:
done

doctoryes commented on tubular/tests/test_retry.py:
```diff
@@ -0,0 +1,125 @@
+        When no max time is set, max_time_reached should always return False
+        """
+        manager = retry.LifecycleManager(1, 1, None)
+        self.assertFalse(manager.max_time_reached())
```
Maybe fake some time passage here and check `max_time_reached()` again?

macdiesel replied:
The code in LifecycleManager.max_time_reached() always returns False if None is passed as the max_time_seconds in the dunder init. I don't think we need anything else here to validate this case. If the variable is set to None, it should return false regardless of the passing of time.

doctoryes commented on tubular/tests/test_retry.py:
```diff
@@ -0,0 +1,125 @@
+
+        # The instance of datetime becomes local to the module it's import in to. We must patch datetime using the
+        # module instance that is imported in to the ec2 module.
+        retry.datetime = NewDateTime
```
🐒

macdiesel replied:
yeah, wish mock.patch would work for this stuff we didn't have to do it manually. =(

doctoryes commented on README.md:
```diff
@@ -12,4 +12,6 @@ Scripts for integrating pipelines with Asgard and ec2 to facilitate Continuous d
 | ASGARD_WAIT_TIMEOUT  | 600                             | Integer - time in seconds to wait for an action such as instances healthy in a load balancer. |
 | REQUESTS_TIMEOUT     | 10                              | How long to wait for an http connection/response from Asgard.                                 |
-
+| RETRY_MAX_ATTEMPTS   | 5                               | Maximum number attempts to be made when asgard returns a 400 or 500 leve response.            |
```
`leve`?

macdiesel replied:
fat fingers... done

doctoryes commented:
First pass complete.

feanil commented on tubular/tests/test_asgard.py:
```diff
@@ -15,6 +17,11 @@
 
 from .test_utils import create_asg_with_tags, create_elb
 
+# Disable the retry decorator and reload the asgard module. This will ensure that tests do not fail because asgard
```
`because asgard`.....?

macdiesel replied:
done

macdiesel commented:
@doctoryes @feanil as to why use sys vs the retry module or an environment variable:
- Environment variable could be used, but it would need to be set before the retry module was loaded. Also since there is no common entry right now for launching tests there isn't a good place to put it. It's easy to specify the variable for travis but what about running the tests locally. I could add a test launching script but then there are docs to add and everyone need to actually read them before executing tests. It just didn't feel user friendly.

As to why in sys and not in the retry module. The variable needs to be set before the retry module is loaded. I needed a module that will not be destroyed during the lifecycle of the application and something that could be shared by all. sys seemed like a good place as any. 

Remember this is only used for testing. If you guys want a longer run down of what works and doesn't work when loading these modules to disable/enable a decorator it might be easier do a 15 minute huddle up and talk about it.

macdiesel commented:
@edx/pipeline-team Ready for another review.

macdiesel commented:
@doctoryes the sys code has been replaced with an environment variable. PTAL.

doctoryes commented:
👍

feanil commented:
👍

timmc-edx committed:
docs: Clean up GoCD alert script docs

- Link to integration that these need to be coordinated with
- Clarify what the auth token is
- Use "JSM" instead of "Opsgenie" everywhere (except to note the old name)
- Fix docstring of close-alert (which had been a copy of open-alert)
- Include alias in open-alert logging
- Add comment next to alias var explaining why it has to be of that form

timmc-edx commented:
This will also allow the scripts to be relatively self-documenting, so that alert runbooks don't all need to explain how the integration works.
